charms/istio-pilot/src/charm.py

```diff
@@ -444,34 +444,37 @@ def _check_leader(self):
             raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
     @property
-    def _ingress_gateway_host(self) -> Optional[str]:
-        """Return a domain name if it's configured, the ingress gateway svc address otherwise."""
-        # If the configuration option is set, we use it
-        if domain_name := self.model.config["domain-name"]:
-            return domain_name
-
-        # We try getting the ingress gateway IP address in the absence of domain name
+    def _cert_subject(self) -> Optional[str]:
+        """Return the certificate subject to be used in the CSR.
+
+        This property takes advantage of the automatic address discovery of the
+        ingress gateway address and uses it to return a name that can be used as
+        a certificate subject when integrated with a TLS certificate provider.
+        The logic is as follows:
+        * Return None if the ingress gateway service address cannot be retrieved
+        * Return a hostname if the address is not an IP (e.g. user-example.com)
+        * Return a hostname if the address is an IP, but there is a hostname that resolves to it
```

Comment:
@DnPlas this is the piece that is hazy to me. How can you detect if a hostname resolves to this IP?

Reply:
But how is this done? How can it know any DNS hostname that could be pointing to this IP?

Reply:
Yeah @kimwnasptd, I had similar questions when @DnPlas and I talked offline today.  I don't have something to point to and say "this is obviously wrong", but this is an area I'm not experienced with so I am not confident in ruling out issues either.

My main worry is this: will `socket.gethostbyaddr()` **always** work, including:
* behind a restrictive proxy
* in an airgapped environment (maybe this matters less - not sure we'd need a CSR at all in an airgapped install?)
My guess is there's probably a use case out there that can break this approach, but I don't know it off hand.  

My secondary worry is whether this might block something even in cases when we don't need a CSR at all.  For example, I'm pretty sure right now if I ran this charm in airgapped the charm will try to execute `socket.gethostbyaddr(LOADBALANCER_IP)` which will raise an exception because it cannot resolve a host.  We can fix that, but my worry is that *we don't have tests right now that catch errors like that*.  So to implement this safely, we need to think that stuff through as well

Reply:
Yeah just confirmed this...

Locally, I can do:
```python
import socket
socket.gethostbyaddr('8.8.8.8')
# ('dns.google', [], ['8.8.8.8'])
```

whereas from a lxc container without a network interface, I see:

```
socket.gethostbyaddr('8.8.8.8')
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# socket.herror: [Errno 2] Host name lookup failure
```

And this code gets called on every `__init__` of the charm, regardless of whether we actually need a CSR

Surely we can work around this, I'm just more worried about the edge cases we haven't thought about yet.

Reply:
fwiw, I like controlling this via config rather than automation because:
1. I don't have a sense of where else this might fail
2. this is not a setting I expect people to constantly be getting frustrated by

Specifically for (2), casual users wont need this (they likely don't need a public host and bulletproof certs), and enterprise users wont keep redeploying kubeflow so they'll set this once and move on.  And in a rare case of needing to deploy with this setting a lot, its also controllable from a bundle.yaml file.  

tl;dr: it would be nice to have, but this is a tiny optimization.  the risk vs reward doesn't feel right to me.

Reply:
I can go back to my original implementation of receiving a configuration option, that is not a problem, that way we could just avoid the edge cases.